### DISCUSSION TITLE
Flow step to execute SQL statement

### DISCUSSION THREAD

**Discussion by terafirmanz** [directus/directus#19862] [Feature Requests]
### Summary

Like the current script block but rather than execute javascript code it would execute SQL code (or send it to the database to be executed)

### Basic Example

I have directus pointed to a database hosted on PostGIS. when I update certain fields in Directus I want a PostGIS function to run to calculate geometry on that data. One example of this is having a latitude/longitude field that PostGIS converts to a spacial point via St_SetPoint(longitude, latitude, 4326).

Currently, this has to be done by manually creating triggers in the database but this means there is no visibility in the admin app of these.

This would not be limited to PostGIS but would open the ability to use stored procedures etc. that may hold more complex functions or business logic or even ETL functions.

### Motivation

To give users direct SQL access from Directus on event/cron without having to write javascript to do the same thing at a higher level. It also provides access to database functions that are not possible or very difficult in JavaScript.

Most of all it allows for job scheduling for the database within the Directus ecosystem. This also helps avoid a circular problem where a flow is triggered to change data that triggers a database trigger that may create an unexpected outcome for a user as they have no visibility as to how the final result came about after the flow finished.

### Detailed Design

same as the flow script action block but rather than executing JavaScript it executes SQL. a function would have to be created taht parsed the SQL before execution and replaced any Directus placeholders with their value

### Must Have:
- A the ability to execute SQL

### Should Have:
- C The ability to pass in Directus values like $trigger.payload.key

### Could Have:
- D The ability to select existing stored procedures found on the DB server

### Drawbacks

Users who use it will have to know both SQL and how to find relationship details within the app to build their queries. As this is only in flows it does not impact users other than those that want to use it.

This would expose the underlying database to users who could execute code to break things including Directus itself. Users have access to their data anyway so could do this directly on the database server.

### Alternatives

Using database triggers.
This moves the admin of the function and its trigger out of Directus and into the database requiring separate administration. 

### Adoption Strategy

not breaking and would be available for anyone who wishes to use it.

### Unresolved Questions

what level of security would we want to put around this if any.

**Comment by rijkvanzanten** [comment]
There's no way for us to do this in a way that's anywhere close to secure :C With the run script operation, we're able to run the javascript in an isolated sandbox that prevents any access to the network, fs, or database etc to prevent malicious code from being able to do anything nefarious. There's no such thing for SQL commands, so I'm very hesitant to entertain the idea of opening up that level of direct access. I think a custom operation would be my recommend solution to this use case.

**Reply by terafirmanz** [reply]
What if it was limited to only executing existing stored procedures. I would have to supply the procedure name then I can pass in variables either direct as values or using Directus variables e.g. $trigger.payload etc.

That way Directus is not running insecure as it can only run functions that exist on the DB server already. A bit of a hybrid but it does allow Directus to be in charge of scheduling and variables based on events a DB may never see. 

Now I can create a complex web of triggers and "dirty" fields in tables to have things happen but I still can't trigger them with cron (in some DB systems) nor can they be triggered via webhook or other things Directus can be.

**Reply by rijkvanzanten** [reply]
That would indeed solve the security problem 🙂 Seeing that we're aiming to support more than just sql, I'm a little hesitant to commit to having a very sql-specific operation included in core, but this would be a perfect extension for in the upcoming marketplace

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
